File: sdk/python/tokenomics_sdk/context.py
"""Ambient attribution context.

Attribution is set once at configure time (project, environment) and refined per call
site with :func:`track`, which nests -- so a request handler can set ``feature`` while an
inner helper adds ``prompt_version`` without either knowing about the other.
"""

from __future__ import annotations

from collections.abc import Iterator
from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field, replace

from tokenomics_sdk.attributes import (
    TOKENOMICS_ENVIRONMENT,
    TOKENOMICS_FEATURE,
    TOKENOMICS_PROJECT,
    TOKENOMICS_PROMPT_VERSION,
    TOKENOMICS_SUBJECT,
    TOKENOMICS_TAG_PREFIX,
)


@dataclass(frozen=True, slots=True)
class Attribution:
    project: str | None = None
    feature: str | None = None
    environment: str | None = None
    subject_id: str | None = None
    prompt_version: str | None = None
    tags: dict[str, str] = field(default_factory=dict)

    def merge(self, other: Attribution) -> Attribution:
        """Overlay non-``None`` fields of ``other``; tags union with ``other`` winning."""
        return Attribution(
            project=other.project or self.project,
            feature=other.feature or self.feature,
            environment=other.environment or self.environment,
            subject_id=other.subject_id or self.subject_id,
            prompt_version=other.prompt_version or self.prompt_version,
            tags={**self.tags, **other.tags},
        )

    def as_attributes(self) -> dict[str, str]:
        mapping = {
            TOKENOMICS_PROJECT: self.project,
            TOKENOMICS_FEATURE: self.feature,
            TOKENOMICS_ENVIRONMENT: self.environment,
            TOKENOMICS_SUBJECT: self.subject_id,
            TOKENOMICS_PROMPT_VERSION: self.prompt_version,
        }
        attributes = {key: value for key, value in mapping.items() if value is not None}
        attributes.update(
            {f"{TOKENOMICS_TAG_PREFIX}{key}": value for key, value in self.tags.items()}
        )
        return attributes


_BASE = Attribution()
# Default is None rather than a shared Attribution instance: the dataclass carries a
# mutable ``tags`` dict, and one shared default could leak tags between contexts.
_CURRENT: ContextVar[Attribution | None] = ContextVar("tokenomics_attribution", default=None)


def set_base(attribution: Attribution) -> None:
    """Set process-wide defaults (called by :func:`tokenomics_sdk.configure`)."""
    global _BASE
    _BASE = attribution


def current() -> Attribution:
    return _BASE.merge(_CURRENT.get() or Attribution())


@contextmanager
def track(
    *,
    project: str | None = None,
    feature: str | None = None,
    environment: str | None = None,
    subject_id: str | None = None,
    prompt_version: str | None = None,
    **tags: str,
) -> Iterator[Attribution]:
    """Attribute every LLM call made inside this block.

    Nests: inner values override outer ones, everything else is inherited.
    """
    overlay = Attribution(
        project=project,
        feature=feature,
        environment=environment,
        subject_id=subject_id,
        prompt_version=prompt_version,
        tags={key: str(value) for key, value in tags.items()},
    )
    merged = (_CURRENT.get() or Attribution()).merge(overlay)
    token = _CURRENT.set(merged)
    try:
        yield merged
    finally:
        _CURRENT.reset(token)


def with_tags(**tags: str) -> Attribution:
    return replace(current(), tags={**current().tags, **tags})

File: sdk/typescript/src/context.ts
/**
 * Ambient attribution context.
 *
 * Attribution is set once at configure time (project, environment) and refined per
 * call site with `track`, which nests via `AsyncLocalStorage` -- so a request handler
 * can set `feature` while an inner `await`-ed helper adds `promptVersion` without
 * either knowing about the other, and concurrent requests never bleed into each other.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import {
  TOKENOMICS_ENVIRONMENT,
  TOKENOMICS_FEATURE,
  TOKENOMICS_PROJECT,
  TOKENOMICS_PROMPT_VERSION,
  TOKENOMICS_SUBJECT,
  TOKENOMICS_TAG_PREFIX,
} from "./attributes.js";

export interface Attribution {
  readonly project?: string;
  readonly feature?: string;
  readonly environment?: string;
  readonly subjectId?: string;
  readonly promptVersion?: string;
  readonly tags: Readonly<Record<string, string>>;
}

export function attribution(partial: Partial<Attribution> = {}): Attribution {
  return {
    project: partial.project,
    feature: partial.feature,
    environment: partial.environment,
    subjectId: partial.subjectId,
    promptVersion: partial.promptVersion,
    tags: partial.tags ?? {},
  };
}

/** Overlay the non-undefined fields of `overlay`; tags union with `overlay` winning. */
export function mergeAttribution(base: Attribution, overlay: Attribution): Attribution {
  return {
    project: overlay.project ?? base.project,
    feature: overlay.feature ?? base.feature,
    environment: overlay.environment ?? base.environment,
    subjectId: overlay.subjectId ?? base.subjectId,
    promptVersion: overlay.promptVersion ?? base.promptVersion,
    tags: { ...base.tags, ...overlay.tags },
  };
}

export function attributionAttributes(value: Attribution): Record<string, string> {
  const mapping: Record<string, string | undefined> = {
    [TOKENOMICS_PROJECT]: value.project,
    [TOKENOMICS_FEATURE]: value.feature,
    [TOKENOMICS_ENVIRONMENT]: value.environment,
    [TOKENOMICS_SUBJECT]: value.subjectId,
    [TOKENOMICS_PROMPT_VERSION]: value.promptVersion,
  };
  const attributes: Record<string, string> = {};
  for (const [key, val] of Object.entries(mapping)) {
    if (val !== undefined) attributes[key] = val;
  }
  for (const [key, val] of Object.entries(value.tags)) {
    attributes[`${TOKENOMICS_TAG_PREFIX}${key}`] = val;
  }
  return attributes;
}

let base: Attribution = attribution();
const storage = new AsyncLocalStorage<Attribution>();

/** Set process-wide defaults (called by {@link configure}). */
export function setBase(value: Attribution): void {
  base = value;
}

export function current(): Attribution {
  return mergeAttribution(base, storage.getStore() ?? attribution());
}

export interface TrackOptions {
  project?: string;
  feature?: string;
  environment?: string;
  subjectId?: string;
  promptVersion?: string;
  tags?: Record<string, string>;
}

/**
 * Attribute every LLM call made inside `fn`. Nests: inner values override outer
 * ones, everything else is inherited. Works across `await` because it rides
 * `AsyncLocalStorage`, not a plain variable.
 */
export function track<T>(options: TrackOptions, fn: () => T): T {
  const overlay = attribution({
    project: options.project,
    feature: options.feature,
    environment: options.environment,
    subjectId: options.subjectId,
    promptVersion: options.promptVersion,
    tags: options.tags,
  });
  const merged = mergeAttribution(storage.getStore() ?? attribution(), overlay);
  return storage.run(merged, fn);
}
